### PR TITLE
Fix double dep path substitution.

### PR DESCRIPTION
**Pull request by goodov**
Fix optimize_webui.py logic to not generate wrong paths for dep files.
First, chromium code do `_get_dep_path` call and then we do same thing on already processed paths. I moved our processing code to `_get_dep_path` to handle our specific paths and return our fixed paths.

Resolves https://github.com/brave/brave-browser/issues/16224

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by goodov** [comment]
I haven't encountered any matching path in Debug build, so I'm not sure whether this code actually does something or not. @mariospr maybe it's matched only in Release or some kind of fully-optimized build?
https://github.com/brave/brave-core/commit/8b1e61f81f4407addc588ac08b19b9df9a14bc3f

**Review by iefremov — APPROVED** [review]

**Comment by mariospr** [comment]
> I haven't encountered any matching path in Debug build, so I'm not sure whether this code actually does something or not. @mariospr maybe it's matched only in Release or some kind of fully-optimized build?

Right, this problem won't manifest itself on a Debug build because it needs to be an optimized one. FWIW, I just checked out your PR, removed some generated files that were presenting the problem before (e.g. `out/Component/gen/chrome/browser/resources/new_tab_page/optimized_js.d`) and rebuilt with `npm run build`, but unfortunately I still see the double path in there:
```
out/Component/gen/chrome/browser/resources/new_tab_page/optimized_js.d:gen/chrome/browser/resources/new_tab_page/new_tab_page.rollup.js: ../../brave/ui/brave/ui/webui/resources/gen/ui/webui/resources/preprocessed/js/assert.m.js ../../brave/ui/brave/ui/webui/resources/gen/chrome/browser/resources/new_tab_page/preprocessed/utils.js[...]
```

... so it seems the PR does not fix this particular issue :neutral_face:

**Comment by goodov** [comment]
> ... so it seems the PR does not fix this particular issue

are you sure you did everything right? did you do `npm run sync`? the issue was in double-called `_get_dep_path`. the modified patch in this PR removes second call, so it should work without introducing `brave/ui.../brave/ui` twice. please double check the final `optimize_webui.py` file in your repo to see if the patch is properly applied :)

**Comment by mariospr** [comment]
> > ... so it seems the PR does not fix this particular issue
> 
> are you sure you did everything right? did you do `npm run sync`? the issue was in double-called `_get_dep_path`. the modified patch in this PR removes second call, so it should work without introducing `brave/ui.../brave/ui` twice. please double check the final `optimize_webui.py` file in your repo to see if the patch is properly applied :)
> 
> this is what I see: `gen\chrome\browser\resources\new_tab_page\new_tab_page.rollup.js: gen\ui\webui\resources\preprocessed\js\assert.m.js gen\chrome\browser\resources\new_tab_page\preprocessed\utils.js gen\chrome\browser\resources\new_tab_page\preprocessed\window_proxy.js`

Oops! Sorry for the confusion, you're absolutely right: I was missing the `npm run apply_patches` step after checking out your branch and now that I run it correctly I can see the paths fixed, thanks!

**Review by mariospr — APPROVED** [review]
LGTM, thanks for fixing this rebasing mistake of mine!

File: patches/chrome-browser-resources-tools-optimize_webui.py.patch [hunk 1]
```diff
@@ -10,11 +10,12 @@ index f41a89bf816fe8cdba73d3700399ce55e94a7e99..8381a1b5be429ef32861c2e9ea36655f
...
-                      request_list)
-+  for (redirect_url, file_path) in [ ('chrome://brave-resources/', _BR_RESOURCES_PATH), ('//brave-resources/', _BR_RESOURCES_PATH) ]: request_list = map(lambda dep: _get_dep_path(dep, redirect_url, file_path), request_list)
- 
-   deps = map(os.path.normpath, request_list)
+@@ -50,6 +51,8 @@ def _request_list_path(out_path, host_url):
+   return os.path.join(out_path, host + '_requestlist.txt')
  
+ def _get_dep_path(dep, host_url, in_path):
++  for brave_host_url in ['chrome://brave-resources/', '//brave-resources/']:
++    if dep.startswith(brave_host_url): return dep.replace(brave_host_url, os.path.relpath(_BR_RESOURCES_PATH, _CWD))
```
Comment (mkarolin) [inline patches/chrome-browser-resources-tools-optimize_webui.py.patch:18]: Could this be squashed into a single line?

Comment (goodov) [reply patches/chrome-browser-resources-tools-optimize_webui.py.patch:18]: I tried that, but no luck. also tried to insert semicolons in some places but still doesn't work.
```
for brave_host_url in ['chrome://brave-resources/', '//brave-resources/']: if dep.startswith(brave_host_url): return dep.replace(brave_host_url, os.path.relpath(_BR_RESOURCES_PATH, _CWD))
                                                                                ^
SyntaxError: invalid syntax
```

Comment (mkarolin) [reply patches/chrome-browser-resources-tools-optimize_webui.py.patch:18]: Ok, thanks for trying.

**Review by mkarolin — APPROVED** [review]
++

**Pull request merged**